modules/manifold/src/utils/default-states.js

```diff
@@ -0,0 +1,279 @@
+import assert from 'assert';
+import {
+  FEATURE_TYPE,
+  FILTER_TYPE,
+  FIELD_ROLE,
+} from '@mlvis/mlvis-common/constants';
+import {dotRange} from '@mlvis/mlvis-common/utils';
+import {product} from './utils';
+
+//---------BaseCols----------//
+
+export const defaultBaseCols = state => {
+  const {columnTypeRanges} = state;
+  assert(
+    columnTypeRanges.score && columnTypeRanges.score.length == 2,
+    '`columnTypeRanges.score` must contain 2 elements'
+  );
+  return dotRange(...columnTypeRanges.score);
+};
+
+export const isValidBaseCols = state => {
+  const {
+    data: {fields},
+    baseCols,
+  } = state;
+  assert(fields.length, '`data.fields` must be non-empty');
+  return (
+    baseCols &&
+    baseCols.length &&
+    baseCols.every(col => 0 <= col && col < fields.length)
+  );
+};
+
+//---------NClusters----------//
+
+export const defaultNClusters = state => {
+  const {
+    data: {columns},
+  } = state;
+  assert(
+    columns.length && columns[0].length,
+    '`data.columns` must be non-empty'
+  );
+  return Math.min(4, columns[0].length);
+};
+
+export const isValidNClusters = state => {
+  const {
+    nClusters,
+    data: {columns},
+  } = state;
+  assert(
+    columns.length && columns[0].length,
+    '`data.columns` must be non-empty'
+  );
+  return (
+    !isNaN(nClusters) && nClusters !== null && nClusters <= columns[0].length
+  );
+};
+
+//---------SegmentFilters----------//
+
+/**
+ * @return {Array<Array<Object>} length corresponds to number of segments, each element is an array of filters each segment must fulfill
+ */
+export const defaultSegmentFilters = state => {
+  const {
+    data: {fields},
+    baseCols,
+  } = state;
+  assert(
+    fields.length && baseCols.length,
+    '`data.field` and `baseCols` must be non-empty'
+  );
+  const baseColFields = baseCols.map(colId => fields[colId]);
+  const filtersByField = baseColFields.map(field =>
+    defaultSegmentFiltersFromFieldDef(field)
+  );
+  // todo: this is not the most efficient way to do filtering (should organize filters by fields). Keeping the structure to for backward compatibility
+  return product(filtersByField);
+};
+
+export const isValidSegmentFilters = state => {
+  const {
+    data: {fields},
+    baseCols,
+    segmentFilters,
+  } = state;
+  assert(
+    fields.length && baseCols.length,
+    '`data.field` and `baseCols` must be non-empty'
+  );
+  return (
+    segmentFilters.length &&
+    segmentFilters.every(segmentFilter => {
+      return (
+        segmentFilter.length &&
+        // each baseCol must have a corresponding filter
+        segmentFilter.length === baseCols.length &&
+        segmentFilter.every((filter, i) => {
+          const colId = filter.key;
+          // each baseCol and their corresponding filter must be in the same order
```

Comment:
Besides returning false, shall we provide (e.g. console.debug or assert) detailed info like this?
Easier to debug.